File: Data Structures & Algorithms/combinations-of-a-phone-number/submission-3.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if len(digits) == 0:
            return []
            
        keyPad = ["", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"]

        res = []

        combLen = len(digits)
        def dfs(i, cur):
            if i == combLen:
                res.append(cur)
                return
            
            for j in range(len(keyPad[int(digits[i])])):
                dfs(i+1, cur + keyPad[int(digits[i])][j])

            
        dfs(0, "")

        return res

File: Data Structures & Algorithms/palindrome-partitioning/submission-3.java
class Solution {
    public List<List<String>> partition(String s) {
        List<List<String>> res = new ArrayList<>();
        List<String> cur = new ArrayList<>();

        dfs(s, cur, 0, res);
        
        return res;
    }

    private void dfs(String s, List<String> cur, int i, List<List<String>> res) {
        if (s.length() == 0) {
            res.add(new ArrayList<>(cur));
            return;
        }

        for (int size = 1; size <= s.length(); size++) {
            if (isPalindrome(s.substring(0, size))) {
                cur.add(s.substring(0, size));
                dfs(s.substring(size), cur, i+size, res);
                cur.remove(cur.size() - 1);
            }
        }
    }

    private boolean isPalindrome(String s) {
        if (s.length() < 2) {
            return true;
        }

        int l = 0;
        int r = s.length() - 1;

        while (l < r) {
            if (s.charAt(l) != s.charAt(r)) {
                return false;
            }
            l++;
            r--;
        }

        return true;
    }
}
